### ISSUE TITLE
Self-overlapping way created by the IHM algorithm 

### ISSUE THREAD

**Issue by zstadler** [IsraelHikingMap/Site#553]
## Bug
[Way 524115341](https://www.openstreetmap.org/way/524115341/history) was created by the IHM algorithm and is self-overlapping.

![image](https://user-images.githubusercontent.com/1304610/32725584-4f706abe-c87e-11e7-93de-12041c7bde54.png)

See [OSM inspector](http://tools.geofabrik.de/osmi/?view=geometry&lon=35.24506&lat=31.98523&zoom=18&opacity=0.67&overlays=self_intersection_ways,self_intersection_points)

**Comment by zstadler** [comment]
Also, [way 519099042](https://www.openstreetmap.org/way/5190990420)
![image](https://user-images.githubusercontent.com/1304610/32765924-5210f4d6-c915-11e7-9b6d-1d9453898bb1.png)

See on [OSM inspector](http://tools.geofabrik.de/osmi/?view=geometry&lon=35.12828&lat=31.93437&zoom=18&opacity=0.67&overlays=self_intersection_ways,self_intersection_points)

**Comment by HarelM** [comment]
@moshfeu Can you please upload the route from @zstadler first example so I'll try to debug this locally?
https://www.openstreetmap.org/api/0.6/gpx/2500207/data

**Comment by moshfeu** [comment]
Is on osm already, I just changed the permission so you can see it here:

https://www.openstreetmap.org/user/Mosh%20Feu/traces/2500207

**Comment by HarelM** [comment]
@zstadler I know this is a bit late but do you think you can find another example?
Both example were fixed and I'm having a hard time understanding the state of the highways at the time of editing.
I was able to load an old pbf file but I'm not sure it's 100% accurate:
The following is the image I managed to create of the scenario:
Green - existing highways.
Blue - Recorded trail.
Red - Algorithm's suggestion.
Black arrow point to a location the where the algorithm created a loop.
![image](https://user-images.githubusercontent.com/3269297/33578414-1fc8f1c8-d94f-11e7-960c-13aa41a5c869.png)
@moshfeu do you remember how you decided to act in terms of adding all missing trails, order of adding, etc?

**Comment by moshfeu** [comment]
@HarelM I'm not sure what you mean.. But the usual way I add trails is to use the algorithm (almost always, then I go to osm editor and fix the ways, or adding new, when the algorithm does not want to add them.. (because they close to an existing way, for example)

**Comment by zstadler** [comment]
@HarelM wrote:
> @zstadler I know this is a bit late but do you think you can find another example?

Currently, there are none. I will send you if there will be any.

> I was able to load an old pbf file but I'm not sure it's 100% accurate

Please let me know if you want a more accurate pbf file

**Comment by HarelM** [comment]
@zstadler it seems that there's a problem debugging the algorithm as it relies on data fetched from OSM when a user want to add a route to OSM after the suggestions (this is to make sure the latest data the algorithm uses is up-to-date).
Due to that fact it'll take some time to be able to fully reproduce the flow that created the issue.
Which made me think that the algorithm may fail due to discrepancy between the data in the database and the data in OSM.
What might solve this issue and allow a better ability to debug is to add the missing data to the database and relay mainly on that.
Which bring us back to #557 - if the ES database will be up-to-date in terms of minutes it might be easier to debug this issue and also improve the performance a bit.
But I'm not sure this is the right course of action...

**Comment by zstadler** [comment]
I didn't understand hoe outdated DB can cause a creation of self-overlapping ways.

Specifically, according to [achavi](https://overpass-api.de/achavi/?changeset=51974897&zoom=18&lat=31.98542&lon=35.24588&layers=B0TTTFT#map=13#map=16) the existing green highway was changed about 3 months prior to the creation of the self-intersection

![image](https://user-images.githubusercontent.com/1304610/33786272-22b04182-dc71-11e7-9f49-b2340ae58363.png)

In any case, having up-to-date DB cannot hurt.

**Comment by zstadler** [comment]
By the way, the order in which the ways were added is:

![image](https://user-images.githubusercontent.com/1304610/33787358-84a02e6c-dc75-11e7-83cb-59dbfbcee10d.png)

1. https://overpass-api.de/achavi/?changeset=51974868
2. https://overpass-api.de/achavi/?changeset=51974893
3. https://overpass-api.de/achavi/?changeset=51974897

**Comment by HarelM** [comment]
Sorry, I wasn't clear, there's no connection between the "outdated" DB and this issue (it's just something I encountered while trying to debug this issue).
Thanks for the extra info! this is exactly what I wanted to understand...

**Comment by HarelM** [comment]
@zstadler do you think you can create a pbf file of israel after changeset 51974893 was made but before 51974897?
I'm having a really hard time creating this pbf for some reason...

**Comment by HarelM** [comment]
@zstadler never-mind, I think I managed to do it using OsmSharp.

**Comment by HarelM** [comment]
Ok, finally got to the bottom of this, I'll use the following image to explain. One thing to bare in mind is that the purpose of `OsmLineAdderService` is to add lines using exiting OSM nodes and create extra nodes when needed.
![image](https://user-images.githubusercontent.com/3269297/34232888-32c5d6a8-e5eb-11e7-80b0-9d7c0767a79f.png)
The state just before adding this line can be seen in the image.
Green line are exiting lines, green points are points on those lines.
Red line is the line we want to add including its points in red.
So the algorithm does the following:
go to first point - 1 - find closest point on closest exiting line and use it.
on second point - 2 - find closes point on closest line according to tolerance - still point 1, since it is used point 2 is ignored.
check if there's an intersection between 2 and 4 with an exiting line - there is - 3 - use it **bug** - we use 1 and 3 which I think are on two consecutive points on the same line.
use 4 since it's not close to anything.
check if there's an intersection between 4 and 5 with an exiting line - there is - 3 - use it **bug**
use point 5 since it is on another line.

Bottom line:
1. There' might be an issue with the missing routes simplification as this small line might not be a correct and relevant addition.
2. The line adder service should have some more checks in order to avoid such issues.

**Comment by zstadler** [comment]
Excellent detective work!

My initial thought is that the problem is in the "create candidate" segments" phase.

![image](https://user-images.githubusercontent.com/1304610/34244285-459e5286-e62d-11e7-9c3d-4153039f0bc1.png)

Segments 2 and 3 above do not seem to follow the "minimal distance between segments" rule. As a result, there is an issue at the "connect new segments to existing ways" stage.

Specifically,

![image](https://user-images.githubusercontent.com/1304610/34244528-4d1d5ac4-e62e-11e7-863d-0c6e3ee98254.png)

Red points 1, 2 and 3 seem to be very close to the north "2" line. Points 1 and 3 are probably overlapping it. Points 2 and 5 also look close to one another.

**Comment by HarelM** [comment]
"minimal distance between segments" is only applicable between exiting segments and new segments.
Segment 2 seems like a good candidate.
While the segment 3 might not be a good candidate since it is short and creates a loop the rest is working as expected in term of how we defined the algorithm.
Bare in mind that what I have debugged in terms of `OsmLineAdder` is the state after the user already added two lines and so this is out side the scope of the candidates algorithm.
Segment 2 is allowed to be close to segment 3 - it is even part of the algorithm to try and join candidates as much as possible in order not to create gap in OSM.

**Comment by HarelM** [comment]
Investigation continues:
I have removed all the non-interesting lines.
The algorithm does the following:
1. FindMissingLines
2. SplitSelfLoopsAndRemoveDuplication
3. SimplifyLines
And get the following results (which I believe are decent, left line is after right line in terms of GPS recording time)
![image](https://user-images.githubusercontent.com/3269297/34293119-1952ece8-e70c-11e7-940d-fe7aa3400593.png)
4. ProlongLinesEndAccordingToOriginalGpx gives the following results which is still acceptable:
![image](https://user-images.githubusercontent.com/3269297/34293185-7477074e-e70c-11e7-9e93-4b5cac9f2b79.png)
5. ProlongLinesStartAccordingToOriginalGpx basically generates the issue but is doing what it should - prolongs the line starting position to join other lines.
![image](https://user-images.githubusercontent.com/3269297/34293266-d8a8f8b2-e70c-11e7-87cf-1eb3393129dc.png)

So I'm guessing the question here is what to do in the following case which is interesting.
My initial thought was to look for pairs of lines which creates loops between themeselfs and check the loop's length - in case it is a small look (say less than 50 meters) remove it and adjust the lines accordingly.

Regardless, the issue with creating an incorrect line within the lineadderservice should be fixed as I was able to think of a scenario that can cause this issue which can't be defined as a problem with the candidates algorithm like in this case.

**Comment by HarelM** [comment]
Thanks to @zstadler  help I was able to improve the splitting function a bit and get the following results:
![image](https://user-images.githubusercontent.com/3269297/34306226-c9077526-e74a-11e7-9ab4-7add0d1be39b.png)
Still need to improve the line prolong function a bit.

**Comment by HarelM** [comment]
After a rewrite of the prolong algorithm the following are the results which I think are far superior to previous results:
![image](https://user-images.githubusercontent.com/3269297/34450505-59c64ed4-ed13-11e7-9251-c2f3f1472453.png)
 The following is a very impressive example of the capabilities of the algorithm - the entire brown line is merged into a single line with an option to add the small segment.
![image](https://user-images.githubusercontent.com/3269297/34450511-97a23ad8-ed13-11e7-8727-f5dc67cd0df0.png)

**Comment by HarelM** [comment]
I've written a wiki page on how the algorithm works in case any of you are interested.
https://github.com/IsraelHikingMap/Site/wiki/Candidates-algorithm:-find-unmapped-routes
